Update destroyer attack behaviour for artillery wagons

Jayefuu opened the pull request.

Jayefuu committed:
Update events.lua

Jayefuu committed:
Update events.lua

grilledham commented on map_gen/maps/crash_site/events.lua:
```diff
@@ -294,6 +294,20 @@ local bot_cause_whitelist = {
+        if not destroyer or not destroyer.valid or not entity or not entity.valid then
+            return
+        end
+        data.entity.surface.create_entity{name = "laser", position=destroyer.position, target=entity, speed=1}
```
Maybe we should add a range check before creating the laser. Seems unfair to the player that the bot can shot the wagon before you have an opportunity to shot it back.

Jayefuu replied:
Done! Thought I'd try be smart and use laser-beam instead which can have a range limit but it looked terrible so I added a range check function instead.

grilledham commented on map_gen/maps/crash_site/events.lua:
```diff
@@ -294,6 +294,20 @@ local bot_cause_whitelist = {
+        if not destroyer or not destroyer.valid or not entity or not entity.valid then
+            return
+        end
+        data.entity.surface.create_entity{name = "laser", position=destroyer.position, target=entity, speed=1}
```
```suggestion
        entity.surface.create_entity{name = "laser", position=destroyer.position, target=entity, speed=1}
```

grilledham commented on map_gen/maps/crash_site/events.lua:
```diff
@@ -294,6 +294,20 @@ local bot_cause_whitelist = {
+            return
+        end
+        data.entity.surface.create_entity{name = "laser", position=destroyer.position, target=entity, speed=1}
+        set_timeout_in_ticks(30, destroyer_callback, {destroyer = destroyer, entity = entity})
```
```suggestion
        set_timeout_in_ticks(30, destroyer_callback, data)
```

Jayefuu committed:
Updated as per suggestions

- Added function to check distance between two entities as I couldn't find one in utils
- Added call to check how far from destroyer to wagon
- Since there's now a check I removed the 2 second wait between spawning and starting to fire in favour of using distance instead (384, 386)

Jayefuu commented:
I think that's everything @grilledham

grilledham commented on map_gen/maps/crash_site/events.lua:
```diff
@@ -294,6 +294,28 @@ local bot_cause_whitelist = {
+        if not destroyer or not destroyer.valid or not entity or not entity.valid then
+            return
+        end
+        if euclidean_distance(destroyer.position, entity.position) < 10 then
```
`math.sqrt` is an expensive calculation, what we have done in the past is return the distance squared and compared that instead.

So effectively something like
```lua
 ((p.x-q.x)^2 + (p.y-q.y)^2) < 100
```
And add a comment saying something like 100 is a distance of 10 squared.

Jayefuu replied:
Smart! I like it. Done.

Jayefuu committed:
Update events.lua
